test/hitl/hitl_test.py

```diff
@@ -0,0 +1,208 @@
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+
+# udev rules :
+# ACTION=="add", SUBSYSTEM=="tty", SUBSYSTEMS=="usb", MODE="0666", PROGRAM="/bin/sh -c 'echo $$ID_SERIAL_SHORT | rev | cut -c -8 | rev'", SYMLINK+="ttyUSB_%c.%s{bInterfaceNumber}"
+# ACTION=="add", SUBSYSTEM=="block", SUBSYSTEMS=="usb", ENV{ID_FS_USAGE}=="filesystem", MODE="0666", PROGRAM="/bin/sh -c 'echo $$ID_SERIAL_SHORT | rev | cut -c -8 | rev'", RUN{program}+="/usr/bin/systemd-mount --no-block --automount=yes --collect $devnode /media/blkUSB_%c.%s{bInterfaceNumber}"
+
+import os
+import sys
+import time
+import serial
+import subprocess
+import json
+
+def flash_jlink(sn, dev, firmware):
+    script = ['halt', 'r', f'loadfile {firmware}', 'r', 'go', 'exit']
+    f = open('flash.jlink', 'w')
+    f.writelines(f'{s}\n' for s in script)
+    f.close()
+    ret = subprocess.run(f'JLinkExe -USB {sn} -device {dev} -if swd -JTAGConf -1,-1 -speed auto -NoGui 1 -ExitOnError 1 -CommandFile flash.jlink',
+                         shell=True, stdout=subprocess.PIPE, stderr=subprocess.STDOUT)
+    stdout = ret.stdout.decode()
+    os.remove('flash.jlink')
+    assert ret.returncode == 0, 'Flash failed\n' + stdout
+
+def test_cdc_dual_ports(id):
+    port1 = f'/dev/ttyUSB_{id[-8:]}.00'
+    port2 = f'/dev/ttyUSB_{id[-8:]}.02'
+    # Wait device enum
+    timeout = 10
+    while timeout:
+        if os.path.exists(port1) and os.path.exists(port2):
+            break
+        time.sleep(1)
+        timeout = timeout - 1
+
+    assert timeout, 'Device not available'
+
+    # Echo test
+    ser1 = serial.Serial(port1)
+    ser2 = serial.Serial(port2)
+
+    ser1.timeout = 1
+    ser2.timeout = 1
+
+    str1 = b"test_no1"
+    ser1.write(str1)
+    ser1.flush()
+    assert ser1.read(100) == str1.lower(), 'Port1 wrong data'
+    assert ser2.read(100) == str1.upper(), 'Port2 wrong data'
+
+    str2 = b"test_no2"
+    ser2.write(str2)
+    ser2.flush()
+    assert ser1.read(100) == str2.lower(), 'Port1 wrong data'
+    assert ser2.read(100) == str2.upper(), 'Port2 wrong data'
+
+    print('cdc_dual_ports test done')
+
+def test_cdc_msc(id):
+    port  = f'/dev/ttyUSB_{id[-8:]}.00'
+    file = f'/media/blkUSB_{id[-8:]}.02/README.TXT'
```

Comment:
@HiFiPhile why would you limit the serial number to 8 characters, wouldn't be simpler to use it as it is ?

Reply:
For some historical reason my udev script cut serial number to 8 characters.

In fact udev rule is not even needed, cdc can be accessed by `/dev/serial/by-id/usb-TinyUSB_TinyUSB_Device_41003B000E504E5457323020-if00`

Reply:
yeah, I figure that out as well. I have already made the changes, just want double check if you first. The id is also shorten in blk device, if it only impact serial, then we are probably safe to remove the cut ?

```python
def get_serial_dev(id, product, ifnum):
    # get usb serial by id
    return f'/dev/serial/by-id/usb-TinyUSB_{product}_{id}-if{ifnum:02d}'


def get_disk_dev(id, lun):
    # get usb disk by id
    return f'/dev/disk/by-id/usb-TinyUSB_Mass_Storage_{id}-0:{lun}'


def get_hid_dev(id, product, event):
    return f'/dev/input/by-id/usb-TinyUSB_{product}_{id}-{event}'

```

Reply:
> The id is also shorten in blk device, if it only impact serial, then we are probably safe to remove the cut ?

Yes you are right.

Reply:
since I am in the middle of this, let me try to do this as well. At worst, we could revert if I messed up :)